README
San simera is a KDE plasmoid which shows (in Greek) the events of the current day back in the history,
from the site www.sansimera.gr

http://kde-apps.org/content/show.php/San+simera?content=156739

To install the plasmoid, run

plasmapkg -i /path/to/sansimera

then add the widget in KDE. If you want to remove it again, run

plasmapkg -r sansimera

Requirements:
    - python-Beautifulsoup4
    - python-pillow (or python-image)

For development, you can also use plasmoidviewer to quickly start the widget from
command line. This way you don't have to install/uninstall all the time.

ChangeLog
Version 1.7.6
- Fix purge old fetched html data
- Adjust space in 'Did you know'

Version 1.7.5
- The event year is displayed in the title to improve arrangement

Version 1.7.4
- Remove previous displaying missing modules (regression in images displaying)

Version 1.7.3
- Add b.C. if needed in years
- Don't crash if a requirement is missing, instead display a message with the needed python package

Version 1.7.2
- Add exception for python-pillow | python-image

Version 1.7.1
- Apply the fonts colour to links
- If offline, try to fetch data if connexion goes online

Version 1.7
- Rewrite code for website changes
- Add icons and links
- Add more events (Nameday, "Did you know", Anniversaries, Quotes)

Version 1.6.2
- Better arrangement of text (alignement from the top)
- Config dialog to change the fonts to black

Version 1.6.1
- Remove whitespaces from text (from birth data also)

Version 1.6
- Remove whitespaces from text

Version 1.5
- Improve code to avoid occupy plasma desktop while switching the events. 
  Now it retrieves the events once on startup. 
  If there is no connection it tries automatically after 20 minutes or when clicking the icon
 
Version 1.4
- By accident i uploaded the version with black coloured text that i used on my laptop.

Version 1.3
- Fix stop working after changes on the website of www.sansimera.gr

Version 1.2
- Fix the crash if there is no connexion  (regression coming from the new features in version 1.1)

Version 1.1
- Added birth and death events

Version 1.0
- Instead of showing all the events at once, it shows, with a time interval, one event the time
- When the icon is clicked it shows the next event of the day

Version 0.3.1
- Don't crash if there is no connection

Version 0.3
- Show the month's name
- Don't define the dimensions of the plasmoid (remembers resizing)

Version 0.2
- Fix encoding problem
- Change the disposition of the title to fit better when resizing in different dimensions
- Don't show the '&nbsp;' in some text

Version 0.1
- First script release
- Show the historical events of the current day

contents/code/main.py
# -*- coding: utf-8 -*-
# Copyright (C) 2013  Dimitrios Glentadakis <dglent@free.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from PyQt4.QtCore import *
from PyQt4.QtGui import *
from PyKDE4.plasma import *
from PyKDE4 import plasmascript
from PyQt4 import uic
from PyKDE4.kdecore import *
from PyKDE4.kdeui import *
import sansimera_data
import sansimera_fetch
import re
import datetime



class Sansimera(plasmascript.Applet):
    def __init__(self,parent,args=None):
        plasmascript.Applet.__init__(self,parent)

    def init(self):
        self.fonts = self.config().readEntry("blackfonts", False).toBool()
        self.setHasConfigurationInterface(True)
        self.download()
        self.data = sansimera_data.Sansimera_data()
        self.lista = self.data.getAll()
        self.index = 0
        self.timer = QTimer(self)
        self.timer1 = QTimer(self)
        self.setMinimumSize(325.0,125.0)
        self.setAspectRatioMode(Plasma.IgnoreAspectRatio)
        self.label = Plasma.TextBrowser(self.applet)
        #self.label = Plasma.Label(self.applet)
        self.icon = Plasma.IconWidget(self.applet)
        self.icon.setIcon(self.package().path() + "contents/icons/sansimera.png")
        self.icon_path = (self.package().path() + "contents/icons/sansimera.png")
        self.connect(self.icon, SIGNAL("clicked()"), self.next_item)
        firstTitleInit = sansimera_fetch.Sansimera_fetch()
        firstTitle = firstTitleInit.monthname()
        self.fetchDate = firstTitleInit.fetchDate()
        self.label.setText(firstTitle)
        self.layout = QGraphicsLinearLayout(Qt.Horizontal, self.applet)
        self.layout.addItem(self.label)
        self.setLayout(self.layout)
        self.san_text()
        QObject.connect(self.timer, SIGNAL("timeout()"), self.san_text)
        self.timer.start(40000)
        QObject.connect(self.timer1, SIGNAL("timeout()"), self.connection_next_try)
        self.timer1.start(1200000)

    def download(self):
        fetch = sansimera_fetch.Sansimera_fetch()
        self.html = fetch.html()
        self.online = fetch.online

    def next_item(self):
        self.timer.start(40000)
        cDate = self.currentDate()
        if self.online == False or cDate != self.fetchDate:
            self.download()
            self.data = sansimera_data.Sansimera_data()
            self.lista = self.data.getAll()
        self.san_text()

    def connection_next_try(self):
        if self.online == False:
            self.download()
            self.data = sansimera_data.Sansimera_data()
            self.lista = self.data.getAll()

    def san_text(self):
        length = len(self.lista)
        self.san_lista = self.lista[self.index]
        self.san_lista = self.trUtf8(self.san_lista)
        self.apply_settings()
        self.label.setText(self.san_lista)
        #self.label.setStyleSheet("background:white; font-weight:normal; color:black;")
        self.index+=1
        if self.index == length:
            self.index = 0

    def createConfigurationInterface(self, parent):
        self.general_widget = QWidget(parent)
        parent.okClicked.connect(self.ConfigAccepted)
        parent.applyClicked.connect(self.ConfigAccepted)
        parent.cancelClicked.connect(self.configWidgetDestroyed)
        self.general_ui = uic.loadUi(self.package().filePath('ui','general.ui'), self.general_widget)
        self.general_ui.blackfonts_checkBox.setChecked(self.config().readEntry("blackfonts", False ).toBool())
        self.general_ui.blackfonts_checkBox.stateChanged.connect(parent.settingsModified)
        parent.addPage(self.general_widget, i18n('General'), self.icon_path)

    def ConfigAccepted(self):
        self.config().writeEntry("blackfonts", bool(self.general_ui.blackfonts_checkBox.isChecked()))
        self.fonts = self.general_ui.blackfonts_checkBox.isChecked()
        self.apply_settings()

    def configWidgetDestroyed(self):
        self.general_widget = None
        self.general_ui = None

    def apply_settings(self):
        if self.fonts:
            self.label.setStyleSheet("color:black;")
            links = re.findall('href="http://[a-zA-Z.0-9/]+">', self.san_lista)
            # Remove links color if present
            linksColor = re.findall('<font color=[a-z]+>', self.san_lista)
            if len(linksColor) > 0:
                for tag in linksColor:
                    self.san_lista = self.san_lista.replace(tag, '')
            # Apply black color for links
            if len(links) > 0:
                for link in links:
                    self.san_lista = self.san_lista.replace(link, link+'<font color=black>')
            self.label.setText(self.san_lista)

        # Apply white color
        if not self.fonts:
            self.label.setStyleSheet("color:white;")
            links = re.findall('href="http://[a-zA-Z.0-9/]+">', self.san_lista)
            # Remove links color if present
            linksColor = re.findall('<font color=[a-z]+>', self.san_lista)
            if len(linksColor) > 0:
                for tag in linksColor:
                    self.san_lista = self.san_lista.replace(tag, '')
            # Apply white color for links
            if len(links) > 0:
                for link in links:
                    self.san_lista = self.san_lista.replace(link, link+'<font color=white>')
            self.label.setText(self.san_lista)

    def currentDate(self):
        cDate = str(datetime.date.today())
        return cDate

def CreateApplet(parent):
    return Sansimera(parent)


contents/code/sansimera_fetch.py
# -*- coding: utf-8 -*-
# Copyright (C) 2013  Dimitrios Glentadakis <dglent@free.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import os
import datetime
import subprocess

class Sansimera_fetch(object):
    def __init__(self):
        self.online = False

    def url(self):
        imerominia = str(self.pay()+self.ponth())
        self.url = 'http://www.sansimera.gr/almanac/'+imerominia
        return self.url

    def pay(self):
        imera = str(datetime.date.today()).split('-')
        imera = imera[1:]
        self.pay = imera[1]
        return self.pay

    def ponth(self):
        imera = str(datetime.date.today()).split('-')
        imera = imera[1:]
        self.ponth = imera[0]
        return self.ponth

    def monthname(self):
        dico = {'01': 'Ιανουαρίου', '02': 'Φεβρουαρίου', '03': 'Μαρτίου', \
            '04': 'Απριλίου', '05': 'Μαίου', '06': 'Ιουνίου', '07': 'Ιουλίου', \
                '08': 'Αυγούστου', '09': 'Σεπτεμβρίου', '10': 'Οκτωβρίου', \
                    '11': 'Νοεμβρίου', '12': 'Δεκεμβρίου'}
        n = self.ponth()
        self.im = str(' '*10+'...Σαν σήμερα '+self.pay()+' '+dico[n]+'\n')
        return self.im

    def html(self):
        cmd = ['kde4-config', '--localprefix']
        output = subprocess.Popen( cmd, stdout=subprocess.PIPE ).communicate()[0]
        path = output[:-1]+'share/apps/plasma/plasmoids/sansimera'
        os.chdir(path)
        link = self.url()
        onoma='sansimera_html'
        comm = 'wget --timeout=5 --user-agent="KDE Plasmoid" '+link+' -O '+onoma
        self.online = True
        os.system(comm)
        arxeio = open(onoma, 'r')
        ss=arxeio.read()
        if ss == '':
            self.online=False
        arxeio.close()

    def fetchDate(self):
        date = str(datetime.date.today())
        return date

if __name__ == "__main__":
    a1=Sansimera_fetch()
    lista=a1.html()



contents/code/sansimera_data_gen.py
# -*- coding:utf-8 -*-
# Copyright (C) 2013  Dimitrios Glentadakis <dglent@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


import sansimera_data


class Sansimera_data_gen(sansimera_data.Sansimera_data):
    def __init__(self):
        self.gen_elist = []


    def parser_gen(self,i):
        ifind = i.index('>')
        i = i[ifind+1:]
        i = i.lstrip()
        if i.count('[Αποφθέγματα]') >= 1:
            i = i.replace("[Αποφθέγματα]", " ")
        if i.count('&nbsp;') >= 1:
            i = i.replace("&nbsp;", " ")
        if len(i) <= 6:
            if i[:-2].isdigit():
                i = '\n'+i # Add the title gennisi and a new line below for the event
        if i != '':
            i = i.rstrip()
            self.gen_elist.append(i+' ')


    def genniseis(self):
        self.gen_htmll = ''
        try:
            arxeio = open('sansimera_html','r')
            gen_html = arxeio.read()
            self.gen_htmll = gen_html.split('h2')
            self.gen_htmll = self.gen_htmll[4]
            self.gen_htmll = self.gen_htmll.split('<')
            arxeio.close()
        except:
            self.htmll = 'Eλέγξτε τη σύνδεσή σας.'
            self.gen_elist.append(self.htmll)
        return self.gen_htmll

    def out_genniseis(self):
        start_count = False
        htmll = self.genniseis()
        for i in htmll:
            if i.count('b>') == 1:
                start_count = True
            if start_count == True:
                i = ' '.join(i.split())
                self.parser_gen(i)
            if i.count('/div>') == 1:
                start_count = False

    def lista_gen(self):
        flag_empty = False
        self.out_genniseis()
        lista = self.gen_elist[:]
        count_lista = []
        count = 0
        for i in lista:
            if len(i) <= 7 and i[len(i)-2:] == ': ':
                count_lista.append(count)
            count+=1
        self.gn_elist = []
        item0 = 0
        item1 = 1
        mikos = len(count_lista)
        lmikos = len(lista)
        flag_end = False

        for i in count_lista:
            it0 = count_lista[item0]
            if mikos == item1:
                it1 = lmikos
                flag_end = True
            if flag_end == False:
                it1 = count_lista[item1]
            pedio=''.join(lista[it0:it1])
            self.gn_elist.append(pedio)
            item0+=1
            item1+=1
        if len(self.gn_elist) == 0:
            flag_empty = True
            self.gn_elist = self.gen_elist[:]

        if len(self.gn_elist)!= 0 and flag_empty == False:
            self.gn1_elist = []
            for i in self.gn_elist:
                i = '\nΓέννηση'+i
                self.gn1_elist.append(i)

        if flag_empty == True:
            return self.gn_elist
        if flag_empty == False:
            return self.gn1_elist



if __name__ == "__main__":
    a1=Sansimera_data_gen()
    lista = a1.lista_gen()
    for i in lista:
        print(i)

contents/code/sansimera_data_than.py
# -*- coding: utf-8 -*-
# Copyright (C) 2013  Dimitrios Glentadakis <dglent@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import sansimera_data
import re


class Sansimera_data_than(sansimera_data.Sansimera_data):
    def __init__(self):
        self.than_elist = []

    def parser_than(self,i):
        text = True
        try:
            ifind = i.index('</b>')
        except:
            text=False

        if text == True:
            i = i.lstrip()
            sfind = i.index('</small>')
            i = i.replace(i[sfind:],"")

        i = re.sub('<[^>]+>', '', i)
        if i.count('[Αποφθέγματα]') >= 1:
            i = i.replace("[Αποφθέγματα]", " ")
        if i.count('&nbsp;') >= 1:
            i = i.replace("&nbsp;", " ")
        if i.count('&rsquo;') >= 1:
            i = i.replace("&rsquo;", "'")
        i='\n'+i
        if i != '':
            i = i.rstrip()
            self.than_elist.append('\nΘάνατος'+i+'\n')

    def thanatoi(self):
        try:
            html_file = open('sansimera_html','r')
            than_html = html_file.read()
            self.than_htmll = than_html.split('h2')
            self.than_htmll = self.than_htmll[6]
            self.than_htmll = self.than_htmll.split("contxtad")
            self.than_htmll = self.than_htmll[0]
            self.than_htmll = self.than_htmll.split("<div class='data_row clear fleft'>")
            html_file.close()
        except:
            self.htmll = 'Δεν βρέθηκαν γεγονότα.'
            self.than_elist.append(self.htmll)
        return self.than_htmll

    def out_thanatoi(self):
        start_count = False
        htmll = self.thanatoi()
        count = 0
        keim = ''.join(htmll)
        htmll = keim.split('<b>')
        # Find the text with year
        pattern = '^[0-9]+: </b>'
        regexp = re.compile(pattern)
        for i in htmll:
            result = regexp.search(i)
            if result:
                # Remove whitespaces
                i = ' '.join(i.split())
                # Send the text with the year to the parser
                self.parser_than(i)

        return self.than_elist

if __name__ == "__main__":
    a1=Sansimera_data_than()
    lista = a1.out_thanatoi()
    for i in a1.than_elist:
        print(i)


contents/code/sansimera_data.py
# -*- coding: utf-8 -*-
# Copyright (C) 2013  Dimitrios Glentadakis <dglent@free.fr>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


from urlparse import urljoin
import re
import urllib
import os
import glob
import sansimera_fetch
import subprocess
from bs4 import BeautifulSoup

try:
    import Image
except ImportError:
    from PIL import Image


class Sansimera_data(object):
    def __init__(self):
        self.allList = []
        self.addBC = False
        self.baseurl = 'http://www.sansimera.gr/'
        # Clean files from a previous fetch
        cmd = ['kde4-config', '--localprefix']
        output = subprocess.Popen( cmd, stdout=subprocess.PIPE ).communicate()[0]
        defaultPath = output[:-1]+'share/apps/plasma/plasmoids/sansimera'
        currentPath = os.getcwd()
        if currentPath == defaultPath:
            plasmoidFiles = ('ChangeLog.txt', 'contents', 'README',
                             'metadata.desktop', 'sansimera_html')
            for _file in glob.glob('*'):
                if _file not in plasmoidFiles:
                    os.remove(_file)
        self.fetch = sansimera_fetch.Sansimera_fetch()
        self.month = self.fetch.monthname()
        self.sanTitle = '&nbsp;'*10+self.month
        if os.path.exists('sansimera_html'):
            with open('sansimera_html') as html:
                self.soup = BeautifulSoup(html)

    def getImage(self, text):
        relativeUrls = re.findall('href="(/[a-zA-Z./_0-9-]+)', text)
        yeartoBold = re.findall('(<div>[0-9]+</div>)', text)
        if len(relativeUrls) > 0:
            for relurl in relativeUrls:
                text = text.replace(relurl, self.baseurl[:-1]+relurl)
        if len(yeartoBold) > 0:
            self.year = yeartoBold[0]
            if self.bC:
                self.year = yeartoBold[0].replace('</div>', ' π.Χ.</div>', re.UNICODE)
            self.year = '<b>'+self.year[5:-6]+':</b>'
            text = text.replace(yeartoBold[0], '')
        try:
            iconUrl = re.findall('src="(http://[a-zA-Z./_0-9-]+)', text)[0]
            iconName = os.path.basename(iconUrl)
            urllib.urlretrieve(iconUrl, iconName)
            im = Image.open(iconName)
            newim = im.resize((36,32), Image.ANTIALIAS)
            newim.save(iconName)
            # Convert the url to local name
            newText = text.replace(iconUrl, iconName)
            return newText
        except:
            return text

    def events(self):
        birthDeath = ''
        listd = self.soup.find_all('div')
        count=0
        for div in listd:
            tag = div.get('class')
            if isinstance(tag, list):
                if len(tag) > 1:
                    # Find the Did You Know ...
                    if tag[0] == 'over' and tag[1] == 'mb10':
                        didYouKnow = (str(listd[count]))
                        # Convert url to local path
                        didYouKnow_url_local = self.getImage(didYouKnow)
                        didYouKnow_url_local = didYouKnow_url_local.replace(
                            'ΗΞΕΡΕΣ ΟΤΙ...', '&nbsp;'*10+'ΗΞΕΡΕΣ ΟΤΙ...')
                        didYouKnow_url_local = didYouKnow_url_local.replace('<br/>', '',1)
                        self.allList.append(didYouKnow_url_local)
                    # Find the He Said ...
                    if tag[0] == 'quote' and tag[1] == 'white':
                        said = ('<br/>' + '&nbsp;'*10 + '<b>Είπε:</b>' + str(listd[count]))
                        whoSaid = str(listd[count+3])
                        # Convert url to local path
                        who_url_local = self.getImage(whoSaid)
                        self.allList.append(said + who_url_local)
                if tag[0] == 'timeline-tab-content':
                    event = listd[count].get('id')
                    if event == 'Deaths':
                        event = self.sanTitle
                        birthDeath = '<br/><small><i>Θάνατος</i></small>'
                    elif event == 'Births':
                        event = self.sanTitle
                        birthDeath = '<br/><small><i>Γέννηση</small></i>'
                    else:
                        event = self.sanTitle
                if tag[0] == 'timeline-item' and tag[1] == 'clearfix':
                    eventText = str(listd[count])
                    # Find if before Christ
                    divBC = str(listd[count-1].p)
                    bC = re.findall('<span>[\.π\s\.Χ]+</span>', divBC, re.UNICODE)
                    if len(bC) == 1:
                            self.bC = True
                    else:
                        self.bC = False
                    eventText_url_local = self.getImage(eventText)
                    self.allList.append(str('<br/>' + event+self.year+birthDeath+eventText_url_local))
            count += 1

    def days(self):
        worldlist = [str('&nbsp;'*20 + '<b>Παγκόσμιες Ημέρες</b><br/>')]
        lista = self.soup.find_all('a')
        for tag in lista:
            url = tag.get('href')
            if isinstance(url, str):
                if 'worldays' in url or 'namedays' in url:
                    if 'worldays' in url:
                        day = 'w'
                    elif 'namedays' in url:
                        day = 'n'
                        title = self.sanTitle + '<br/>' + '&nbsp;'*20 + '<b>Εορτολόγιο</b><br/>'
                    tag = str(tag)
                    if 'Εορτολόγιο' in tag or 'Παγκόσμιες Ημέρες' in tag:
                        continue
                    # Parse with getImage() to expand relative urls
                    tag = self.getImage(tag)
                    if day == 'w':
                        worldlist.append('<br/>' + tag + '.')
                    elif day == 'n':
                        tag = '<br>'+title+'<br/>' + tag
                        self.allList.append(tag)
        if len(worldlist) > 1:
            worldays = ' '.join(worldlist)
            worldays = '<b/>'+self.sanTitle + '<br/>' + worldays
            self.allList.append(worldays)

    def getAll(self):
        self.allList = []
        self.events()
        self.days()
        if len(self.allList) == 0:
            self.allList.append('<br/>' + self.sanTitle + '<br/><br/>Δεν βρέθηκαν γεγονότα, ελέγξτε τη σύνδεσή σας.')
        return self.allList

if __name__ == "__main__":
    a1=Sansimera_data()
    lista = a1.getAll()
    for i in lista:
        print(i)
        raw_input()
